---
title: "STACAS vs other integration methods on heterogeneous T cell datasets"
author: "M. Andreatta and S. Carmona"
date: "26/08/2022"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'STACAS.compareMethods.TILs.html'))})
---

# Introduction

Here we will compare integration results of STACAS with other R packages on a collection of scRNA-seq datasets of mouse tumor-infiltrating T cells from multiple studies from [Andreatta et al](https://www.nature.com/articles/s41467-021-23324-4)

The data are available at: [figshare/23136746](https://figshare.com/ndownloader/files/23136746)

Test also https://figshare.com/ndownloader/files/23166794

# R environment

Get and load some useful packages
```{r message=F, warning=F,results=F, eval=T}
renv::restore()

if (!require("remotes", quietly = TRUE))
    install.packages("remotes")
library(remotes)

if (!requireNamespace("STACAS", quietly = TRUE))
  remotes::install_github("carmonalab/STACAS")

if (!requireNamespace("harmony", quietly = TRUE))
  remotes::install_github("immunogenomics/harmony", ref="6866e01") #fix version at Sep 8, 2022

if (!requireNamespace("SeuratWrappers", quietly = TRUE))
remotes::install_github('satijalab/seurat-wrappers')

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

if (!require("batchelor", quietly = TRUE))
BiocManager::install("batchelor")

if (!require("tidyr", quietly = TRUE))
install.packages("tidyr")

if (!require("tidytext", quietly = TRUE))
install.packages("tidytext")

if (!require("scIntegrationMetrics", quietly = TRUE))
install_github("carmonalab/scIntegrationMetrics") #calculates LISI and Silhouette

```


```{r message=F, warning=F,results=F}
library(Seurat)
library(dplyr)
library(ggplot2)
library(STACAS)
library(harmony)
library(SeuratWrappers)
library(batchelor)
library(tidyr)
library(scIntegrationMetrics)
library(patchwork)
library(tidytext)

seed = 1234
set.seed(seed)
```


# Load test datasets

A merged dataset is available as a Seurat object
```{r}
download <- F
where <- 'aux'
dir.create(where, showWarnings = FALSE)

rds.path <- sprintf("%s/ref_TILAtlas_mouse_v1.rds", where)

if(download){
  options(timeout=500)
  url <- "https://figshare.com/ndownloader/files/23136746"
  download.file(url = url, destfile = rds.path)
}

object <- readRDS(rds.path)
DefaultAssay(object) <- "RNA"
```

Annotated subtypes and dataset/study are stored in `functional.cluster` and `Study` metadata column, respectively.

```{r}
meta.batch <- "Study"
meta.label <- "functional.cluster"
```


```{r}
batchLabels.table <- table(object@meta.data[[meta.label]], object@meta.data[[meta.batch]])
batchLabels.table
```

For integration metrics, do not consider labels that are contributed in more than `thrMaxPerBatch` (90%) by a single dataset/batch.
Additionally, at least `minBatchesPerCellType` (2) datasets have to contribute with at least `thrMinPerBatch` (5%) each. Then number of datasets over this threshold can be used to calculate normalized LISI scores per label.

```{r}
batchLabels <- round( batchLabels.table / rowSums(batchLabels.table) * 100) 
batchLabels
thrMaxPerBatch <- 90
minBatchesPerCellType <- 2
thrMinPerBatch <- 5
```

```{r}
bachesPerLabel <- apply(batchLabels,1,function(x) sum(x > thrMinPerBatch) )
removeLabels <- names(bachesPerLabel[bachesPerLabel < minBatchesPerCellType ])
removeLabels <- unique(removeLabels, 
                       names(which(apply(
                         batchLabels, 1, function(x) max(x) > thrMaxPerBatch) 
                         ))) #  contributed in more than 90% by a single dataset/batch
removeLabels
metricsLabels <- setdiff(unique(object@meta.data[[meta.label]]),removeLabels)
```



Set critical parameters
```{r}
nfeatures <- 2000 # number of highly variable genes for dimensionality reduction
ndim <- 20 # number of PCA components for dimensionality reduction 
lisi_perplexity <- 30 # number of neighbors for LISI (see below for details) to measure batch mixing
```


How does the collection of datasets look without any integration?

Run a standard Seurat pipeline for dimensionality reduction.


```{r, eval=F}
library(data.table)

#The table of standard genes in Ensembl can be download from [this link](https://github.com/carmonalab/scRNAseq_data_processing/blob/master/aux/EnsemblGenes105_Mmu_GRCm39.txt.gz)


ensembleRef <- "aux/EnsemblGenes105_Mmu_GRCm39.txt.gz"

object@assays$RNA@counts <- object@assays$RNA@data

object <- STACAS:::standardizeGeneSymbols(obj = object, EnsemblGeneFile=ensembleRef)
```

```{r message=F, warning=F,results=F}
object <- FindVariableFeatures(object, nfeatures = nfeatures) %>%
  ScaleData() %>%
  RunPCA(npcs=ndim) %>%
  RunUMAP(dims=1:ndim)
```

```{r, results=F, warning=F, message=F }
integrationMetrics <- list()

useMetrics <- c("batch_LISI","batch_nLISI", "batch_nLISI_perCellType", "batch_nLISI_perCellType_means", 
"1-celltype_nLISI", "1-celltype_nLISI_means", "celltype_ASW", "celltype_ASW_means")

method = "uncorrected"
method.reduction <- "pca"
metricsObject <- object

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)

```

Set up plot parameters
```{r}
plot.list <- list()
metricsShow.batch <- grep("batch",names(integrationMetrics[[1]]),value = T)
metricsShow.celltype <- grep("celltype",names(integrationMetrics[[1]]),value = T)
```

```{r}

plotIntegrationMetricsDim <- function(plot.list, metricsShow.batch, metricsShow.celltype, reduction="umap", plot=T){

metricsShow.batch.caption <- paste(metricsShow.batch,round(as.numeric(integrationMetrics[[method]][metricsShow.batch]),2),collapse = "\n ")
metricsShow.celltype.caption <- paste(metricsShow.celltype,round(as.numeric(integrationMetrics[[method]][metricsShow.celltype]),2),collapse = "\n ")
  
plot.list[[method]][["batch"]] <- DimPlot(metricsObject, group.by = meta.batch, reduction = reduction) + theme(aspect.ratio = 1) + labs(subtitle = "Dataset/batch", title = method, caption = metricsShow.batch.caption)

plot.list[[method]][["label"]] <- DimPlot(metricsObject, group.by = meta.label, label=T, label.size = 4, reduction = reduction) + theme(aspect.ratio = 1) + labs(subtitle = "Cell labels", title = method, caption = metricsShow.celltype.caption)


if(plot) print(plot.list[[method]][["batch"]] | plot.list[[method]][["label"]])

return(plot.list)

} 

```

```{r fig.height=8, fig.width=8}

plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```

Clusters are largely driven by study/batch of origin rather than subtype


A simple approach to mitigate batch effects is to select HVGs that are consistently variable across datasets.
For this we'll split by dataset/batch, calculate HVG for each, then identify shared genes using Seurat's SelectIntegrationFeatures.


```{r}
obj.list <- SplitObject(object, split.by = meta.batch)
```

```{r message=F, warning=F,results=F, eval=T}

for (i in 1:length(obj.list)) {
  #obj.list[[i]] <- NormalizeData(obj.list[[i]],assay="RNA",normalization.method="LogNormalize") # these data are already pre-normalized
  obj.list[[i]] <- FindVariableFeatures(obj.list[[i]], nfeatures=nfeatures*2)
}

hvg <- SelectIntegrationFeatures(obj.list, nfeatures = nfeatures)
```


Alternatively, calculate HVG excluding specific gene sets using FindVariableFeatures.STACAS function (default STACAS' behaviour)
```{r warning=F, collapse =T,message =F, eval=F}
library(SignatuR)
my.genes.blocklist <- GetSignature(SignatuR$Hs)

for (i in 1:length(obj.list)) {
  #obj.list[[i]] <- NormalizeData(obj.list[[i]],assay="RNA",normalization.method="LogNormalize")
  obj.list[[i]] <- FindVariableFeatures.STACAS(obj.list[[i]], nfeat=nfeatures*2, genesBlockList = my.genes.blocklist)
}

hvg <- SelectIntegrationFeatures(obj.list, nfeatures = nfeatures)
```


Re-calculate dimensionality reduction using `hvg`
```{r message=F, warning=F,results=F}
object@assays$RNA@var.features <- hvg
object <- ScaleData(object) %>% RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```

```{r message=F, warning=F,results=F} 
method = "sharedFeatures"
method.reduction <- "pca"
metricsObject <- object

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)


```

```{r fig.height=8, fig.width=8}
plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```


# STACAS integration

```{r, message=F, warning=F,results=F}
object_integrated_stacas <- Run.STACAS(obj.list, dims = 1:ndim, anchor.features = hvg) %>%
      RunUMAP(dims = 1:ndim) 
```

```{r, results=F, warning=F, message=F}
method = "STACAS"
method.reduction <- "pca"
metricsObject <- object_integrated_stacas

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)

```

```{r fig.height=8, fig.width=8}
plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```



# Semi-supervised STACAS integration

When available, cell type annotations can be used to guide the alignment. STACAS will use this information to penalize anchors where cell types are inconsistent.
In this dataset, cells were annotated by the authors of the benchmark. In practice we will partially annotate cells using tools such as [scGate](https://github.com/carmonalab/scGate)

Here we indicate in `cell.labels` the metadata column that contains cell annotations
```{r, results=F, warning=F, message=F }
object_integrated_ss <- obj.list %>% Run.STACAS(dims = 1:ndim, anchor.features = hvg, cell.labels = meta.label)
```


Note that there is no need for ALL cells to be annotated: we recommend to set labels to *NA* or *unknown* for cells that cannot be confidently annotated, and they won't be penalized for label inconsistency. In addition, you can decide how much weight to give to cell labels with the `label.confidence` parameter (from 0 to 1).

Visualize on UMAP space
```{r results=F, warning=F, message=F}
object_integrated_ss <- object_integrated_ss %>% RunUMAP(dims=1:ndim)
```

```{r, results=F, warning=F, message=F}
method = "semisupSTACAS"
method.reduction <- "pca"
metricsObject <- object_integrated_ss

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)


```

```{r fig.height=8, fig.width=8}
plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```



# Seurat Integration - CCA method

```{r, results=F, warning=F, message=F, fig.height=8, fig.width=16, eval=T}

# find anchors
obj.list.seuratAnchors <- FindIntegrationAnchors(obj.list, anchor.features = hvg, reduction = "cca", dims = 1:ndim)

# integrate data
object_integrated_cca <- IntegrateData(anchorset = obj.list.seuratAnchors, dims=1:ndim)

rm(obj.list.seuratAnchors)


#Visualize on UMAP space

object_integrated_cca <- object_integrated_cca %>% ScaleData() %>% RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)

method = "SEURAT_cca"
method.reduction <- "pca"
metricsObject <- object_integrated_cca

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)

plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```



# Seurat Integration - RPCA method

```{r, results=F, warning=F, message=F }

# find anchors
obj.list.seuratAnchors <- FindIntegrationAnchors(obj.list, anchor.features = hvg, reduction = "rpca", dims = 1:ndim)

# integrate data
object_integrated_rpca <- IntegrateData(anchorset = obj.list.seuratAnchors, dims=1:ndim)

rm(obj.list.seuratAnchors)
```

Visualize on UMAP space
```{r results=F, warning=F, message=F}

object_integrated_rpca <- object_integrated_rpca %>% ScaleData() %>% RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```

```{r, results=F, warning=F, message=F}
method = "SEURAT_rpca"
method.reduction <- "pca"
metricsObject <- object_integrated_rpca

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)

```

```{r fig.height=8, fig.width=8, eval=T}
plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```




# Harmony Integration

```{r, message=F, warning=F,results=F}
object_integrated_harmony <- RunHarmony(object, group.by.vars =  meta.batch) 
```

Visualize on UMAP space
```{r results=F, warning=F, message=F}
object_integrated_harmony <- RunUMAP(object_integrated_harmony, reduction = "harmony", dims=1:ndim)
```


```{r, results=F, warning=F, message=F}
method = "Harmony"
method.reduction <- "harmony"
metricsObject <- object_integrated_harmony

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)

```

```{r fig.height=8, fig.width=8, eval=T}
plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```


# FastMNN Integration

```{r, message=F, warning=F,results=F}
object_integrated_FastMNN <- RunFastMNN(object.list = obj.list, features = hvg, assay="RNA", d=ndim)
```

Visualize on UMAP space
```{r results=F, warning=F, message=F}
object_integrated_FastMNN <- RunUMAP(object_integrated_FastMNN, reduction = "mnn", dims=1:ndim)
```


```{r, results=F, warning=F, message=F}
method = "FastMNN"
method.reduction <- "mnn"
metricsObject <- object_integrated_FastMNN

integrationMetrics[[method]] <- getIntegrationMetrics(object=metricsObject, metrics = useMetrics, meta.label = meta.label, meta.batch = meta.batch, lisi_perplexity = lisi_perplexity, method.reduction = method.reduction, metricsLabels = metricsLabels)


```

```{r fig.height=8, fig.width=8, eval=T}
plot.list <- plotIntegrationMetricsDim(plot.list,metricsShow.batch,metricsShow.celltype)

```



# Summary of Integration Metrics


```{r fig.height=15, fig.width=15}

integrationMetricsSummary <- data.frame(unlist(integrationMetrics)) %>% tibble::rownames_to_column() %>% dplyr::rename(value=unlist.integrationMetrics.) %>% separate(rowname, c("Method","Metric"), sep="\\.") #\\.[[:alpha:]]

ggplot(integrationMetricsSummary,aes(x=reorder_within(Method,-value, Metric), y=value, fill=Method)) + geom_bar(stat="identity") + 
    theme_bw() + 
    theme(legend.position="none", axis.text.x=element_text(angle = 90, vjust = 0.5, hjust=1)) + xlab("Dataset") + facet_wrap(~Metric, scales = "free")

ggsave("STACAS.compareMetrics.Tcells.summary.png", width = 20, height = 20)
```

```{r}
integrationMetricsSummary %>% filter(Metric %in% c("celltype_ASW_means","batch_nLISI_perCellType_means")) %>% pivot_wider(names_from = Metric, values_from = value ) %>% ggplot(aes(x=batch_nLISI_perCellType_means, y=celltype_ASW_means, label=Method)) + geom_point(aes(color=Method))  + geom_text(hjust=0.1, vjust=0) +  theme_light()
```

```{r  fig.height=20, fig.width=20}
p <- wrap_plots(lapply(plot.list,wrap_plots),ncol=2)
p
ggsave("STACAS.compareMetrics.Tcells.umaps.png",p, width = 30, height = 30)

```


# Further reading

The STACAS package and installation instructions are available at: [STACAS package](https://github.com/carmonalab/STACAS)

The code for this demo can be found on [GitHub](https://github.com/carmonalab/STACAS.demo)

# References

* Andreatta A., Carmona S. J. (2021). *STACAS: Sub-Type Anchor Correction for Alignment in Seurat to integrate single-cell RNA-seq data.* - Bioinformatics

* Luecken, M. D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Müller, M. F., ... & Theis, F. J. (2022). *Benchmarking atlas-level data integration in single-cell genomics.* - Nature methods

* Hao, Y., Hao, S., Andersen-Nissen, E., Mauck III, W. M., Zheng, S., Butler, A., ... & Satija, R. (2021). *Integrated analysis of multimodal single-cell data.* - Cell


