---
title: "Data integration with STACAS"
author: "M. Andreatta and S. Carmona"
date: "29/01/2024"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'STACAS.demo.html'))})
---


STACAS is a method for scRNA-seq integration that is particuarly suited for collections of datasets with large cell type imbalance.

Prior cell type knowledge, given as cell labels, can be provided to the algorithm to perform semi-supervised integration, leading to increased preservation of biological variability in the data. STACAS is robust to partial and imperfect cell type labels and can be applied to large-scale integrations.


In this demo we will show the application of STACAS to integrate a collection of scRNA-seq datasets of immune cells from multiple donors, human tissues and studies, assembled by [Luecken et al.](https://www.nature.com/articles/s41592-021-01336-8) for their comprehensive benchmark.

The data are available at: [figshare/12420968](https://figshare.com/articles/dataset/Benchmarking_atlas-level_data_integration_in_single-cell_genomics_-_integration_task_datasets_Immune_and_pancreas_/12420968)

# R environment

Get and load some useful packages
```{r message=F, warning=F,results=F, eval=T}
renv::restore()

if (!require("remotes", quietly = TRUE))
    install.packages("remotes")
library(remotes)

if (!requireNamespace("STACAS", quietly = TRUE))
  remotes::install_github("carmonalab/STACAS")
```


```{r message=F, warning=F,results=F}
library(Seurat)
library(dplyr)
library(ggplot2)
library(STACAS)

seed = 1234
set.seed(seed)
```

# Load test datasets

Download the dataset of human immune cells assembled by [Luecken et al.](https://www.nature.com/articles/s41592-021-01336-8), and convert them to Seurat objects.
```{r}
download <- FALSE
where <- 'aux'
dir.create(where, showWarnings = FALSE)

rds.path <- sprintf("%s/Immune_ALL_human.rds", where)

if(download){

  options(timeout=5000)
  url <- "https://figshare.com/ndownloader/files/25717328"
  h5.path <- sprintf("%s/Immune_ALL_human.rds", where)
  download.file(url = url, destfile = h5.path)
  
  if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

  if (!require("zellkonverter", quietly = TRUE))
    install.packages("zellkonverter") # to convert from 'h5ad' to 'sce' object
  
  #Convert to Seurat 
  object.sce <- zellkonverter::readH5AD(h5.path)
  object <- Seurat::as.Seurat(object.sce, counts = "counts", data = "X")
  object <- RenameAssays(object, originalexp="RNA")
  rm (object.sce)
  
  Idents(object) <- "final_annotation"
  saveRDS(object = object,file = rds.path)

}else{
  object <- readRDS(rds.path)
}
```

Cell types were annotated by the authors on each dataset individually, using a common dictionary of cell types (see https://github.com/theislab/scib-reproducibility).
These are stored in the `final_annotation` metadata column. Study of origin is stored in `batch` metadaa

```{r}
table(object$final_annotation, object$batch)[,1:5]
```

How does the collection of datasets look without any integration? Run a standard Seurat pipeline for dimensionality reduction and visualization.

```{r message=F, warning=F,results=F}
nfeatures <- 1000
ndim <- 20
object <- FindVariableFeatures(object, nfeatures = nfeatures) %>%
  NormalizeData() %>% ScaleData() %>%
  RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```

```{r fig.height=7, fig.width=16}
p1_pre <- DimPlot(object, group.by = "batch") + theme(aspect.ratio = 1) +
  ggtitle("Dataset/batch before integration")
p2_pre <- DimPlot(object, group.by = "final_annotation", label=T, label.size = 5) +
  NoLegend() +
  theme(aspect.ratio = 1) + ggtitle("Cell labels before integration")


p1_pre | p2_pre
```

Although cells mostly cluster by the cell type (as annotated in individual datasets), there are also visible batch effects (seen as dataset-specific clustering).

# One-liner STACAS

STACAS integration can be performed with a single one-liner command:
```{r, warning=F, collapse =T,message =F}
object_integrated <- object %>% SplitObject(split.by = "batch") %>%
      Run.STACAS(dims = 1:ndim, anchor.features = nfeatures) %>%
      RunUMAP(dims = 1:ndim) 

DimPlot(object_integrated, group.by = "batch")
```

# Step-by-step STACAS integration

For more control over the integration steps in STACAS, you can run individual steps separately and inspect intermediate results.

## 1. Find integration anchors between datasets/batches
```{r, warning=F, collapse =T,message =F}
obj.list <- SplitObject(object, split.by = "batch")

stacas_anchors <- FindAnchors.STACAS(obj.list, 
                                     anchor.features = nfeatures,
                                     dims = 1:ndim)
```

## 2. Guide tree for integration order
```{r,collapse=T,message=F,warning=F}
st1 <- SampleTree.STACAS(
  anchorset = stacas_anchors,
  obj.names = names(obj.list)
  )
```

Are samples from the same sequencing technology or from similar tissues clustering together? Different hierarchical clustering methods are available as `hclust.methods` parameter.

## 3. Dataset integration
```{r results=F, warning=F, message=F}
object_integrated <- IntegrateData.STACAS(stacas_anchors,
                                          sample.tree = st1,
                                          dims=1:ndim)
```

Calculate low-dimensional embeddings and visualize integration results in UMAP
```{r results=F, warning=F, message=F}
object_integrated <- object_integrated %>% ScaleData() %>%
  RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```


```{r fig.height=7, fig.width=16}
p1_int <- DimPlot(object_integrated, group.by = "batch") +
  theme(aspect.ratio = 1) + ggtitle("Dataset/batch after integration")
p2_int <- DimPlot(object_integrated, group.by = "final_annotation", label=T, label.size = 5) +
  NoLegend() + theme(aspect.ratio = 1) + ggtitle("Cell labels after integration") 

p1_int | p2_int
```

At least visually, cells from different studies appear to be better mixed than the uncorrected data. In a later section we will use quantitative metrics to verify whether integration was successful in removing batch effects.

# Semi-supervised integration

When available, cell type annotations can be used to guide the alignment. STACAS will use this information to penalize anchors where cell types are inconsistent.

In this dataset, cells were annotated by the authors of the benchmark. For your own data, you may want to do manual annotation or apply one of several cell annotation tools, such as [SingleR](https://bioconductor.org/packages/release/bioc/html/SingleR.html), [Garnett](https://cole-trapnell-lab.github.io/garnett/) or [scGate](https://github.com/carmonalab/scGate). We will be posting some examples of cell type annotation using `scGate` in a different demo.

Run **semi-supervised STACAS** as one-liner by indicating the metadata column that contains cell annotations (`cell.labels` in this case):
```{r, results=F, warning=F, message=F }
object_integrated_ss <- obj.list %>%
  Run.STACAS(dims = 1:ndim, anchor.features = nfeatures, cell.labels = "final_annotation")
```


Note that there is no need for ALL cells to be annotated: we recommend to set labels to *NA* or *unknown* for cells that cannot be confidently annotated, and they won't be penalized for label inconsistency. In addition, you can decide how much weight to give to cell labels with the `label.confidence` parameter (from 0 to 1).

Visualize on UMAP space
```{r results=F, warning=F, message=F}
object_integrated_ss <- object_integrated_ss %>% RunUMAP(dims=1:ndim)
```

```{r fig.height=7, fig.width=16}
p1_ss <- DimPlot(object_integrated_ss, group.by = "batch") +
  theme(aspect.ratio = 1) +
  ggtitle("Dataset/batch after semi-supervised integration")
p2_ss <- DimPlot(object_integrated_ss, group.by = "final_annotation", label=T, label.size = 5) + 
  NoLegend() + theme(aspect.ratio = 1) + ggtitle("Cell labels after semi-supervised integration")

p1_ss | p2_ss
```


# Metrics of integration quality

Two main aspects should be considered to determine the quality of single-cell data integration: *(i)* batch mixing and *(ii)* preservation of biological variability.

Batch mixing measures whether similar cells originating from different batches are well mixed after integration. A common metrics to quantify batch mixing is the [integration LISI](https://github.com/immunogenomics/LISI); here we will use a celltype-aware version of the integration LISI (CiLISI), to account for different composition between datasets, as implemented in the [scIntegrationMetrics package](https://github.com/carmonalab/scIntegrationMetrics).

Preservation of biological variability can be quantified by how close to each other cells of the same type are, and how separated from each other cells of different types are in the joint integrated embedding. A useful metric for preservation of biological variability is Average Silhouette Width (ASW) of cell labels (celltype_ASW) in the corrected PCA space.

```{r message=F, warning=F, results=F}
library(remotes)
if (!require("scIntegrationMetrics", quietly = TRUE))
install_github("carmonalab/scIntegrationMetrics") #calculates LISI and Silhouette

library(scIntegrationMetrics)

integrationMetrics <- list()
integrationMetrics[["uncorrected"]] <- getIntegrationMetrics(object=object,
                                                      metrics = c("CiLISI","celltype_ASW"),
                                                      meta.label = "final_annotation",
                                                      meta.batch = "batch")

integrationMetrics[["STACAS"]] <- getIntegrationMetrics(object=object_integrated,
                                                      metrics = c("CiLISI","celltype_ASW"),
                                                      meta.label = "final_annotation",
                                                      meta.batch = "batch")

integrationMetrics[["ssSTACAS"]] <- getIntegrationMetrics(object=object_integrated_ss,
                                                      metrics = c("CiLISI","celltype_ASW"),
                                                      meta.label = "final_annotation",
                                                      meta.batch = "batch")

```

```{r warning=FALSE, message=F}
if (!require("tidyr", quietly = TRUE))
install.packages("tidyr")

integrationMetricsSummary <- data.frame(unlist(integrationMetrics)) |>
  tibble::rownames_to_column() |> rename(value=unlist.integrationMetrics.) |>
  separate(rowname, c("Method","Metric"), sep="\\.")

a <- integrationMetricsSummary %>% filter(Metric=="CiLISI") %>%
  ggplot(aes(x=Method, y=value, fill=Method)) + geom_bar(stat="identity") + 
  ggtitle("CiLISI") + theme_bw() +
  theme(legend.position="none", axis.text.x=element_blank())
  

b <- integrationMetricsSummary %>% filter(Metric=="celltype_ASW") %>%
  ggplot(aes(x=Method, y=value, fill=Method)) + geom_bar(stat="identity") + 
  ggtitle("celltype_ASW") + theme_bw() +
  theme(axis.text.x=element_blank()) 
  

a | b
```

Compared to the uncorrected data, both STACAS and ssSTACAS achieve better batch mixing (CiLISI), and improved clustering of cells of the same type (celltype_ASW). Semi-supervised integration mixes slightly less different batches compared to unsupervised STACAS, but is better at preserving biological variability (i.e. higher celltype_ASW).

# Label transfer between datasets

In some integration tasks, cell type annotations may be available only for a fraction of the datasets. In such cases, labels can be "transferred" to unannotated datasets by similarity to annotated cells. The `annotate.by.neighbors` function allows propagating cell labels to unannotated cells by K-nearest neighbor similarity to annotated cells.

For example, if one of the datasets were lacking cell type labels (set to NA):
```{r}
incomplete <- object
incomplete$final_annotation[incomplete$batch == "10X"] <- NA
```

Transfer labels from annotated to non-annotated cells
```{r fig.height=6, fig.width=13}
complete <- annotate.by.neighbors(incomplete, labels.col = "final_annotation")

a <- DimPlot(incomplete, group.by = "final_annotation", label=T, label.size = 5) + NoLegend() +
  theme(aspect.ratio = 1) + ggtitle("Incomplete annotation")

b <- DimPlot(complete, group.by = "final_annotation", label=T, label.size = 5) + NoLegend() +
  theme(aspect.ratio = 1) + ggtitle("Complete annotation")

a | b
```

# Notes on data integration

1. The first critical step in batch effect correction is to ensure that the **gene names in your datasets are harmonized**. If different studies use different naming conventions, synonyms, etc., methods for dimensionality reduction will treat these genes as entirely different variables, and introducing artificial differences between datasets. It is recommended that your pre-processing pipeline takes care of resolving ambiguities in gene naming across datasets. A useful function is `STACAS::StandardizeGeneSymbols()`, but there are also other tools such as [HGNChelper](https://cran.r-project.org/web/packages/HGNChelper/index.html)

```{r eval=F}
#load conversion table
library(STACAS)
data(EnsemblGeneTable.Hs)

#convert gene names separately for each sample
obj.list <- lapply(obj.list, function(x) {
  StandardizeGeneSymbols(x, EnsemblGeneTable = EnsemblGeneTable.Hs)
})
```

2. The calculation of highly variable genes (HVG) is a fundamental step for dimensionality reduction and integration. We recommend excluding certain classes of genes, e.g. mitochondrial, ribosomal, heat-shock genes, from the list of HVG, as their expression are typically more strongly associated to technical variation  than to actual biological differences. The `FindVariableGenes.STACAS()` function allows providing a list of genes to exclude from HVG; see an example below. We can use the collection of signatures stored in [SignatuR package](https://github.com/carmonalab/SignatuR)
```{r eval=T}
if (!require("SignatuR", quietly = TRUE))
    install_github("carmonalab/SignatuR")

library(SignatuR)
#Retrieve full list of signatures for human
hs.sign <- GetSignature(SignatuR$Hs)
```

Then we tell STACAS to exclude specific genes from HVG used in integration, using the parameter `genesBlockList`
```{r warning=F, collapse =T,message =F, eval=F}

my.genes.blocklist <- c(GetSignature(SignatuR$Hs$Blocklists),
                        GetSignature(SignatuR$Hs$Compartments))

object_integrated_blockList <- Run.STACAS(obj.list, genesBlockList = my.genes.blocklist,
                                          dims = 1:ndim, anchor.features = nfeatures)

```

# Further reading

The STACAS package and installation instructions are available at: [STACAS package](https://github.com/carmonalab/STACAS)

The code for this demo can be found on [GitHub](https://github.com/carmonalab/STACAS.demo)

# References

* Andreatta A., Carmona S. J. (2021). *STACAS: Sub-Type Anchor Correction for Alignment in Seurat to integrate single-cell RNA-seq data.* - Bioinformatics

* Andreatta M, Herault L, Gueguen P, Gfeller D, Berenstein AJ, Carmona SJ (2024) *Semi-supervised integration of single-cell transcriptomics data* - Nature Communications

* Luecken, M. D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Müller, M. F., ... & Theis, F. J. (2022). *Benchmarking atlas-level data integration in single-cell genomics.* - Nature methods

* Hao, Y., Hao, S., Andersen-Nissen, E., Mauck III, W. M., Zheng, S., Butler, A., ... & Satija, R. (2021). *Integrated analysis of multimodal single-cell data.* - Cell

# Session info

```{r}
sessionInfo()
```

