---
title: "Data integration with STACAS"
author: "M. Andreatta, A. Berenstein and S. Carmona"
date: "12/04/2022"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'STACAS.demo.html'))})
---

# Introduction

STACAS is a method for scRNA-seq data integration. It is based on the [Seurat](https://cran.r-project.org/web/packages/Seurat/index.html) integration framework, and adds the following innovations:

* **anchors are filtered/down-weighted** based on their distance in reciprocal PCA space, calculated from the unscaled, normalized data
* **integration trees** are constructed based on the 'centrality' of datasets, as measured by the sum of their anchor (distance-weighted) scores
* **cell labels**, if known, can be given as input to the algorithm to perform **semi-supervised integration**

In this demo we will show the application of STACAS to integrate a collection of scRNA-seq datasets of immune cells from multiple donors, human tissues and studies, assembled by [Luecken et al.](https://www.nature.com/articles/s41592-021-01336-8) for their excellent benchmark.

The data are available at: [figshare/12420968](https://figshare.com/articles/dataset/Benchmarking_atlas-level_data_integration_in_single-cell_genomics_-_integration_task_datasets_Immune_and_pancreas_/12420968)

# R environment

Get and load some useful packages
```{r message=F, warning=F,results=F, eval=T}
renv::restore()

if (!require("remotes", quietly = TRUE))
    install.packages("remotes")

library(remotes)

install_github("https://github.com/carmonalab/STACAS")
```


```{r message=F, warning=F,results=F}
library(Seurat)
library(dplyr)
library(ggplot2)
library(STACAS)

seed = 1234
set.seed(seed)
```

# Load test datasets

Download the dataset of human immune cells assembled by [Luecken et al.](https://www.nature.com/articles/s41592-021-01336-8), and convert them to Seurat objects.
```{r}
download <- F
where <- 'aux'
dir.create(where, showWarnings = FALSE)

rds.path <- sprintf("%s/Immune_ALL_human.rds", where)

if(download){

  options(timeout=500)
  url <- "https://figshare.com/ndownloader/files/25717328"
  h5.path <- sprintf("%s/Immune_ALL_human.rds", where)
  download.file(url = url, destfile = h5.path)
  
  if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

  if (!require("zellkonverter", quietly = TRUE))
    install.packages("zellkonverter") # to convert from h5ad to R object
  
  #Convert to Seurat 
  object.sce <- zellkonverter::readH5AD(h5.path)
  object <- Seurat::as.Seurat(object.sce, counts = "counts", data = "X")
  object <- RenameAssays(object, originalexp="RNA")
  rm (object.sce)
  
  Idents(object) <- "final_annotation"
  saveRDS(object = object,file = rds.path)

}else{
  object <- readRDS(rds.path)
}
```

Cell types were annotated by the authors on each dataset individually, using a common dictionary of cell types (see https://github.com/theislab/scib-reproducibility).
These are stored in the `final_annotation` metadata column. Study of origin is stored in `batch` metadaa

```{r}
table(object$final_annotation, object$batch)[,1:5]
```

How does the collection of datasets look without any integration?

Run a standard Seurat pipeline for dimensionality reduction

```{r message=F, warning=F,results=F}
nfeatures <- 2000
ndim <- 50
object <- FindVariableFeatures(object, nfeatures = nfeatures) %>% NormalizeData() %>% ScaleData() %>% RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```

```{r fig.height=7, fig.width=16}
p1_pre <- DimPlot(object, group.by = "batch") + theme(aspect.ratio = 1) + ggtitle("Dataset/batch before integration")
p2_pre <- DimPlot(object, group.by = "final_annotation", label=T, label.size = 4) + theme(aspect.ratio = 1) + ggtitle("Cell labels before integration")


p1_pre | p2_pre
```

Although cells mostly cluster by the cell type (as annotated in individual datasets) but there are also visible batch effects (seen as dataset-specific clustering) 


Let's quantify dataset/batch mixing using Raychaudhuri Lab's [LISI](https://github.com/immunogenomics/LISI), and conservation of biological information by means of Average Silhouette Width (ASW) of cell labels in the corrected PCA space:
```{r}

integrationMetrics <- list()

if (!require("scIntegrationMetrics", quietly = TRUE))
install_github("carmonalab/scIntegrationMetrics") #calculates LISI and Silhouette
library(scIntegrationMetrics)

integrationMetrics[["uncorrected"]] <- list()

integrationMetrics[["uncorrected"]][["batch_LISI"]] <- mean(compute_lisi(object@reductions[["pca"]]@cell.embeddings, meta_data = object@meta.data, label_colnames="batch", perplexity = 20)[[1]]) # perplexity=effective number of each cell's neighbors; use lower than default to make it faster

integrationMetrics[["uncorrected"]][["celltype_ASW"]] <- mean(compute_silhouette(object@reductions[["pca"]]@cell.embeddings, meta_data = object@meta.data, label_colnames="final_annotation")[[1]])

integrationMetrics[["uncorrected"]]

```



Let's run STACAS for correcting batch effects.

# STACAS unsupervised integration

STACAS takes a list of Seurat objects as input, so let's first split the merged object into a list of individual batches/datasets
```{r, warning=F, collapse =T, message =F}
obj.list <- SplitObject(object, split.by = "batch")
```


**STACAS Step 1:** Find integration anchors
```{r, warning=F, collapse =T,message =F}
stacas_anchors <- FindAnchors.STACAS(obj.list, 
                                     anchor.features = nfeatures,
                                     dims = 1:ndim)
```

**STACAS step 2:** dataset integration
```{r results=F, warning=F, message=F}
object_integrated <- IntegrateData.STACAS(stacas_anchors, 
                                              dims=1:ndim)
```

Calculate low-dimensional embeddings and visualize integration results in UMAP
```{r results=F, warning=F, message=F}
object_integrated <- object_integrated %>% ScaleData() %>% RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```


```{r fig.height=7, fig.width=16}
p1_int <- DimPlot(object_integrated, group.by = "batch") + theme(aspect.ratio = 1) + ggtitle("Dataset/batch after integration")
p2_int <- DimPlot(object_integrated, group.by = "final_annotation", label=T, label.size = 2) + theme(aspect.ratio = 1) + ggtitle("Cell labels after integration") 

p1_int | p2_int
```

**Quantify integration metrics**:
```{r}
integrationMetrics[["STACAS"]] <- list()

integrationMetrics[["STACAS"]][["batch_LISI"]] <- mean(compute_lisi(object_integrated@reductions[["pca"]]@cell.embeddings, meta_data = object@meta.data, label_colnames="batch", perplexity = 20)[[1]])

integrationMetrics[["STACAS"]][["celltype_ASW"]] <- mean(compute_silhouette(object_integrated@reductions[["pca"]]@cell.embeddings, meta_data = object@meta.data, label_colnames="final_annotation")[[1]])

integrationMetrics[["STACAS"]]
```


Compared to the non-corrected data, we observed now 
i) increase in dataset/batch LISI (i.e. effective number of different batches in a cell neighborhood), indicating a higher mixing
ii) increase in Silhouette/ASW, indicating that cells with the same annotation (ie of the same kind) were brought into proximity


# One-liner STACAS

The previous integration could have been run with a single command:
```{r, eval=F}
object_integrated <- Run.STACAS(obj.list, dims = 1:ndim, anchor.features = nfeatures)
```


# STACAS integration guide trees

Like Seurat, STACAS uses a guide tree to determine integration order.
Optionally, you can check the integration guide tree automatically generated by STACAS 
(e.g. are samples from the same sequencing technology or from similar tissues clustering together?)

```{r,collapse=T,message=F,warning=F}
st1 <- SampleTree.STACAS(
  anchorset = stacas_anchors,
  obj.names = names(obj.list)
  )
```

You can tune or manually re-calculate this tree and passed it to `IntegrateData.STACAS`:

**STACAS step 2 with pre-calculated integration tree**
```{r results=F, warning=F, message=F, eval=F}
object_integrated <- IntegrateData.STACAS(stacas_anchors, 
                                              dims=1:ndim, 
                                              sample.tree=st1
                                              )
```




# STACAS integration - semi-supervised

When available, cell type annotations can be used to guide the alignment. STACAS will use this information to penalize anchors where cell types are inconsistent.

In this dataset, cells were annotated by the authors of the benchmark. For your own data, you may want to do manual annotation or apply one of several cell annotation tools, such as [SingleR](https://bioconductor.org/packages/release/bioc/html/SingleR.html), [Garnett](https://cole-trapnell-lab.github.io/garnett/) or [scGate](https://github.com/carmonalab/scGate). We will be posting some examples of cell type annotation using `scGate` in a different demo.


**semi-supervised STACAS step 1:** Calculate anchors taking into account cell labels by indicating the corresponding metadata column in `cell.labels` 

```{r, warning=F, collapse =T,message =F}
stacas_anchors_ss <- FindAnchors.STACAS(obj.list,
                                        anchor.features = nfeatures,
                                        dims = 1:ndim,
                                        label.confidence = 1,
                                        cell.labels = "final_annotation")
```

Note that there is no need for ALL cells to be annotated: we recommend to set labels to *NA* or *unknown* for cells that cannot be confidently annotated, and they won't be penalized for label inconsistency. In addition, you can decide how much weight to give to cell labels with the `label.confidence` parameter (from 0 to 1).


Optionally, check out the new default integration tree 
```{r}
st2 <- SampleTree.STACAS(
  anchorset = stacas_anchors_ss,
  obj.names = names(obj.list)
  )
```

**semi-supervised STACAS step 2:** Integrate the datasets using the labels-informed anchor set, using `semisupervised=TRUE`
```{r results=F, warning=F, message=F}
object_integrated_ss <- IntegrateData.STACAS(stacas_anchors_ss, 
                                             dims=1:ndim,
                                             semisupervised=TRUE)
```


** One-liner semi-supervised STACAS **

The previous integration could have been run with the single command:
```{r, eval=F}
object_integrated_ss <- Run.STACAS(obj.list, dims = 1:ndim, anchor.features = nfeatures, cell.labels = "final_annotation" )
```


Calculate low-dimensional embeddings and see integration results in UMAP space
```{r results=F, warning=F, message=F}
object_integrated_ss <- object_integrated_ss %>% ScaleData() %>% RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```

```{r fig.height=7, fig.width=16}
p1_ss <- DimPlot(object_integrated_ss, group.by = "batch") + theme(aspect.ratio = 1) + ggtitle("Dataset/batch after semi-supervised integration")
p2_ss <- DimPlot(object_integrated_ss, group.by = "final_annotation", label=T, label.size = 2) + theme(aspect.ratio = 1) + ggtitle("Cell labels after semi-supervised integration")

p1_ss | p2_ss
```


**Quantify integration metrics**:
```{r}
integrationMetrics[["semisupSTACAS"]] <- list()

integrationMetrics[["semisupSTACAS"]][["batch_LISI"]] <- mean(compute_lisi(object_integrated_ss@reductions[["pca"]]@cell.embeddings, meta_data = object@meta.data, label_colnames="batch", perplexity = 20)[[1]])

integrationMetrics[["semisupSTACAS"]][["celltype_ASW"]] <- mean(compute_silhouette(object_integrated_ss@reductions[["pca"]]@cell.embeddings, meta_data = object@meta.data, label_colnames="final_annotation")[[1]])

integrationMetrics[["semisupSTACAS"]]
```

In this case we observe a >20% increase in cell annotation silhouette (celltype_ASW), with a similar dataset/batch mixing (batch_LISI) compared to the unsupervised integration.



# Summary of Integration Metrics


```{r}

integrationMetricsSummary <- data.frame(unlist(integrationMetrics)) %>% tibble::rownames_to_column() %>% rename(value=unlist.integrationMetrics.) %>% separate(rowname, c("Method","Metric"), sep="\\.")

a <- integrationMetricsSummary %>% filter(Metric=="batch_LISI") %>% ggplot(aes(x=Method, y=value, fill=Method)) + geom_bar(stat="identity") + theme(legend.position="none") + ggtitle("batch_LISI")

b <- integrationMetricsSummary %>% filter(Metric=="celltype_ASW") %>% ggplot(aes(x=Method, y=value, fill=Method)) + geom_bar(stat="identity")+ theme(legend.position="none")+ ggtitle("celltype_ASW")

a | b
```



# Notes on data integration

1. The first critical step in batch effect correction is to ensure that the **gene names in your datasets are harmonized**. If different studies use different naming conventions, synonyms, etc., methods for dimensionality reduction will treat these genes as entirely different variables, and introducing artificial differences between datasets. It is recommended that your pre-processing pipeline takes care of resolving ambiguities in gene naming across datasets. You can use  `STACAS:::standardizeGeneSymbols` but there are other tools such as [HGNChelper](https://cran.r-project.org/web/packages/HGNChelper/index.html)

2. The calculation of highly variable genes (HVG) is a fundamental step for dimensionality reduction and integration. We recommend excluding certain classes of genes, e.g. mitochondrial, ribosomal, heat-shock genes, from the list of HVG, as their expression are typically more strongly associated to technical variation  than to actual biological differences. The `FindVariableGenes.STACAS()` function allows providing a list of genes to exclude from HVG; see an example below.



We can use the collection of signatures stored in [SignatuR package](https://github.com/carmonalab/SignatuR)
```{r eval=T}
if (!require("SignatuR", quietly = TRUE))
    install_github("carmonalab/SignatuR")

library(SignatuR)
print(SignatuR$Hs)

lapply(GetSignature(SignatuR$Hs),head) # retrieve full list of signatures and display 
```

Then we tell STACAS to exclude specific genes from HVG used in integration, using the parameter `genesBlockList`
```{r warning=F, collapse =T,message =F, eval=F}

my.genes.blocklist <- c(GetSignature(SignatuR$Hs$Blocklists),GetSignature(SignatuR$Hs$Cell_types,SignatuR$Hs$Compartments))

object_integrated_blockList <- Run.STACAS(obj.list, genesBlockList = my.genes.blocklist, dims = 1:ndim, anchor.features = nfeatures)

```


# Further reading

The STACAS package and installation instructions are available at: [STACAS package](https://github.com/carmonalab/STACAS)

The code for this demo can be found on [GitHub](https://github.com/carmonalab/STACAS.demo)

# References

* Andreatta A., Carmona S. J. (2021). *STACAS: Sub-Type Anchor Correction for Alignment in Seurat to integrate single-cell RNA-seq data.* - Bioinformatics

* Luecken, M. D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Müller, M. F., ... & Theis, F. J. (2022). *Benchmarking atlas-level data integration in single-cell genomics.* - Nature methods

* Hao, Y., Hao, S., Andersen-Nissen, E., Mauck III, W. M., Zheng, S., Butler, A., ... & Satija, R. (2021). *Integrated analysis of multimodal single-cell data.* - Cell


